text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,107 @@
+
+## Motivation
+
+At present, Yidian Zixun adopts TiKV RawKV for Key-Value storage. Due to performance concerns, TxnKV is not used. However, RawKV only provides simple KV semantics, and does not support multiple versions based on timestamp, which causes some problems: 
+  - When services are migrated from HBase to TiKV, RawKV does not fulfil some major functionality requirements such as data expiration and historical data version;
+  - When developing backup and synchronization functions based on RawKV, it is difficult to solve the problems of online data migration, incremental backup, and data synchronization between two data centers because timestamp is not supported.
+
+In order to solve the above problems, this proposal presents another kv interface - VerKV, which mainly adds version information such as timestamp to the key. At the same time, this interface will be designed to be approximate to the semantics of HBase, which will facilitate the service migration from HBase to TiKV, thereby integrating TiKV better into the HBase ecosystem.
+
+## Detailed designs
+
+### Functions and scenarios
+
+Here are the functionality designs and the corresponding application scenarios for VerKV:
+   - Support data migration, incremental backup, CDC (Change Data Capture) and cross-IDC cluster synchronization application scenarios include:
+     - Online data replication between TiKV clusters, which include a full migrations followed by incremental migrations;
+     - Cluster data backup, which includes a first full backup followed  perform an incremental data backups every other time;
+     - CDC service created based on incremental data acquisition; can reliably provide incremental data services;
+     - In cross data center cluster synchronization scenarios, VerKV can not only process  incremental data synchronization, but also handle write conflicts caused by the same key.
+
+  - Support multi-versioned data. The corresponding application scenarios include: 
+    - Data version retrieval. For certain businesses, multiple versions of historical data must be maintained for for backtracking and analysis. Reliance on third-party storage and consumption of computing resources (user dynamic characteristics , User behavior link tracking) needs to be reduced, and the data processing process optimized;
+    - Data version control, for fast data recovery after problematic data writes in businesses.
+
+  - Support data expiration based on TTL. The corresponding application scenarios include:
+    - Setting expiration time to obtain data of a recent period
+    - Setting expiration time to clear inactive data for a specified amount of time
+
+
+### Compatibility
+
+The design adds a VerKV Column Family in RocksDB, which avoids poential compatibility issues with the existing RawKV annd TxnKV data and processing logics, and could be created dynamically when needed. This design allows maximum reuse of the current implementations of TiKV and PD.
+
```

Comment:
We already have some problem with mixing raw and transactional APIs (as described in the motivation). It seems to me that there is potential to make this worse by adding a third API. As part of this RFC I would like to see some mechanism for better enforcing separation of API use.

Reply:
the problem with mixing raw and transactional APIs is for those two apis use the same `default` cf. so here verkv api use a separated cf. later, it is possible for rawkv to use other cf, then it is ok to use those three apis simultaneously.

Reply:
aiui, that is not the only problem - using the raw API it is possible to edit any CF, which can mean circumventing or just breaking the transactional API. So it is only ok to use both APIs if you don't access any of the three CFs used for transactions with the raw API. There is also back-compat hazard (e.g., if a user has a CF called VerKV already).

As a mechanism for separating APIs, I think separating CFs is fine, I guess we would need some kind of opt-in for users who deliberately want to access VerKV/default/write/lock CFs